lib/Dialect/TritonGPU/Transforms/Pipeliner/SoftwarePipeliner.cpp

```diff
@@ -31,17 +29,27 @@ using namespace mlir;
 #define GEN_PASS_CLASSES
 #include "triton/Dialect/TritonGPU/Transforms/Passes.h.inc"
 
-// The main loop pipelining. Return the new ForOp if it was successfully
-// pipelined.
-static void pipelineLoop(scf::ForOp forOp, int numStages) {
-  mlir::triton::PipeliningOption options;
+// Return true if the preconditions for pipelining the loop are met.
+bool preCondition(scf::ForOp forOp) {
   // Skip loop with distance > 1 for now.
   // TODO: relax the constraint in the expander.
   if (llvm::any_of(forOp.getBody()->getTerminator()->getOperands(),
                    [](Value operand) {
                      Operation *def = operand.getDefiningOp();
                      return !def;
                    }))
+    return false;
+  for (Operation &op : forOp.getBody()->without_terminator()) {
+    // Only pipeline single block loops.
+    if (isa<scf::ForOp, scf::IfOp, scf::WhileOp>(op))
```

Comment:
Why also skip `IfOp`? I can imagine situations where you may have benign `scf.if`s and still pipeline the surrounding for.

Reply:
We currently don't test those cases at any places so it seems odd to support it. I guess IfOp should be okay as long as we check that the nested blocks don't have loops.